### ISSUE TITLE
Problem with SAM2 model

### ISSUE THREAD

**Issue by TasTab007** [ultralytics/ultralytics#19002]
I am trying to use SAM2 from Ultralytics but it's not working. Is SAM2 is being updated? Why it's not working?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TasTab007, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. This will significantly help us understand the problem and replicate the issue.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples, training logs, and details about your environment. Verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Please ensure that you are using the latest version of the `ultralytics` package, along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). Run the following command to update:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response, but please be assured that an Ultralytics engineer will assist you soon. Feel free to provide more details in the meantime, such as any error messages or the steps you followed when encountering the issue. 🚀

**Comment by TasTab007** [comment]
I updated the version to newest one and still getting this error:

![Image](https://github.com/user-attachments/assets/e4521635-c5a3-4fcd-b3bc-e3e6b8177f23)

**Comment by Y-T-G** [comment]
You need to provide the code and the full error traceback

**Comment by TasTab007** [comment]
Here is the code:
import os
import cv2
import torch
import json
import logging
import time
import numpy as np
import matplotlib.pyplot as plt
from ultralytics import SAM
from sklearn.metrics import precision_recall_curve, auc
from pycocotools import mask as mask_utils  # Import pycocotools for RLE decoding

# Set up logging
logging.basicConfig(level=logging.INFO)

# Paths for dataset
DATASET_PATH = r"/home/myid/tt59612/Panoptic Segmentation/Dataset/Coral Net.v8-final-dataset.sam2"
TRAIN_PATH = os.path.join(DATASET_PATH, "train")
VALID_PATH = os.path.join(DATASET_PATH, "valid")
TEST_PATH = os.path.join(DATASET_PATH, "test")

# Pre-trained SAM2 model path
SAM_MODEL_PATH = r"/home/myid/tt59612/Panoptic Segmentation/sam_l.pt"  # Update this path to your SAM2 weights

# Set device (Use GPU if available)
device = "cuda" if torch.cuda.is_available() else "cpu"
logging.info(f"Using device: {device}")

# Output folder for results
RESULTS_DIR = "sam2_fine_tuned_result"
SEGMENTED_IMAGES_DIR = os.path.join(RESULTS_DIR, "segmented_images")
os.makedirs(SEGMENTED_IMAGES_DIR, exist_ok=True)

# Log file for storing evaluation results
EVAL_LOG_FILE = os.path.join(RESULTS_DIR, "evaluation_log.txt")

# Function to decode RLE segmentation masks
def decode_rle(segmentation):
    """Decodes RLE segmentation mask to a binary mask."""
    if isinstance(segmentation, dict) and "counts" in segmentation and "size" in segmentation:
        return mask_utils.decode(segmentation).astype(np.uint8) * 255  # Convert to 0-255 range
    return None

# Function to load images and masks from JSON files
def load_data(image_folder):
    image_files = [f for f in os.listdir(image_folder) if f.endswith((".jpg", ".png", ".jpeg"))]
    data = []
    
    for image_file in image_files:
        image_path = os.path.join(image_folder, image_file)
        json_path = os.path.splitext(image_path)[0] + ".json"

        if not os.path.exists(json_path):
            logging.warning(f"Skipping {image_file}: No corresponding JSON file found.")
            continue

        # Load image
        img = cv2.imread(image_path)
        if img is None:
            logging.warning(f"Skipping {image_file}: Unable to read image.")
            continue

        # Load JSON annotation
        with open(json_path, "r") as f:
            annotations = json.load(f)
        
        # Extract instance segmentation masks
        masks = []
        for ann in annotations.get("annotations", []):
            segmentation = ann.get("segmentation", None)

            if segmentation is None:
                logging.warning(f"Skipping {image_file}: No segmentation data found.")
                continue

            # Handle RLE-encoded masks
            if isinstance(segmentation, dict) and "counts" in segmentation:
                mask = decode_rle(segmentation)
                if mask is not None:
                    masks.append(mask)
                else:
                    logging.warning(f"Skipping {image_file}: Failed to decode RLE mask.")
            
            # Handle polygon segmentation
            elif isinstance(segmentation, list):  
                mask = np.zeros(img.shape[:2], dtype=np.uint8)
                for poly in segmentation:
                    pts = np.array(poly, np.int32).reshape((-1, 2))
                    cv2.fillPoly(mask, [pts], 255)
                masks.append(mask)
            
            else:
                logging.warning(f"Skipping {image_file}: Unknown segmentation format.")

        if len(masks) > 0:
            data.append((img, masks))

    return data

try:
    # Load SAM2 model
    sam_model = SAM(SAM_MODEL_PATH).to(device)

    # Load training and validation data
    logging.info("Loading training data...")
    train_data = load_data(TRAIN_PATH)
    logging.info(f"Loaded {len(train_data)} training samples.")

    logging.info("Loading validation data...")
    valid_data = load_data(VALID_PATH)
    logging.info(f"Loaded {len(valid_data)} validation samples.")

    # Fine-tune SAM2 on the dataset
    logging.info("Starting SAM2 fine-tuning...")
    sam_model.train(data=train_data, val=valid_data, epochs=30, imgsz=650)
    
    # Save the fine-tuned model
    fine_tuned_sam_model = os.path.join(RESULTS_DIR, "fine_tuned_sam2.pt")
    sam_model.save(fine_tuned_sam_model)
    logging.info(f"✅ SAM2 fine-tuning completed. Model saved at {fine_tuned_sam_model}")

    # Apply SAM2 on test images
    logging.info("Applying SAM2 instance segmentation on test images...")

    # Metrics storage
    iou_scores = []
    dice_scores = []
    processing_times = []

    for image_name in os.listdir(TEST_PATH):
        image_path = os.path.join(TEST_PATH, image_name)

        if not image_path.endswith((".jpg", ".png", ".jpeg")):
            continue

        img = cv2.imread(image_path)
        if img is None:
            logging.warning(f"Skipping {image_name}: Unable to read image.")
            continue

        start_time = time.time()
        results = sam_model(img)
        end_time = time.time()

        if isinstance(results, list) and len(results) > 0:
            segmented_image = results[0].masks
        else:
            segmented_image = None

        # Save segmented images
        output_path = os.path.join(SEGMENTED_IMAGES_DIR, image_name)
        if segmented_image is not None:
            cv2.imwrite(output_path, segmented_image)
        else:
            cv2.imwrite(output_path, img)

        # Record processing time
        processing_times.append(end_time - start_time)

    # Compute FPS and Memory Usage
    avg_fps = len(processing_times) / sum(processing_times)
    gpu_memory = torch.cuda.memory_allocated() / (1024 ** 3)

    # Save evaluation log
    with open(EVAL_LOG_FILE, "w") as f:
        f.write("🟢 Evaluation Results:\n")
        f.write(f"📌 FPS (Inference Speed): {avg_fps:.2f} frames/sec\n")
        f.write(f"📌 GPU Memory Usage: {gpu_memory:.2f} GB\n")

    logging.info(f"✅ Segmentation and evaluation completed. Results saved in: {RESULTS_DIR}")

except Exception as e:
    logging.error(f"❌ An error occurred: {str(e)}")

**Comment by Y-T-G** [comment]
You can't train SAM in Ultralytics.

Also follow Ultralytics guide instead of AI-generated code.

https://docs.ultralytics.com/models/sam/
